lib/pbench/server/api/resources/query_apis/controllers_list.py

```diff
@@ -22,20 +22,22 @@ def __init__(self, config: PbenchServerConfig, logger: Logger):
             config,
             logger,
             Schema(
-                Parameter("user", ParamType.USER, required=True),
+                Parameter("user", ParamType.USER, required=False, nullable=True),
```

Comment:
I'm still mentally struggling with the exact semantics I want for user vs access, so my fantasies aren't completely realized at this point and are probably grist for another PR beyond this. Still, a few ideas to mull:

1. My original concept was that missing "user" meant "default to logged-in user" but `"user": ""` and/or `"user": None` meant "published documents". This is largely based on observation that Aquib's dashboard changes currently result in "`"user": ""` when there's no user logged in. One big hole here is that there's no way to specify "I want _my_ data _plus_ all public data";
2. I added "access" to be able to handle that case (although the current code doesn't); e.g., `"user": "dave", "access": "public"` means ... well, then I realized _this_ is ambiguous, too. Do we want to consider a third case of "show me only _my_ data which I've _published_"??? And, if so, how do we express the distinction between "union" vs "intersection"... and are these two parameters the right way to approach that?
3. I'm not sure I see a good case for a distinction between `"user": None` and missing `"user"` (that is, the `nullable=True`) now that I added `access`... except that it's what the dashboard currently does. (It'd be slightly more complicated for the dashboard to omit the parameter when there's no username.) While making things easy on the dashboard is a good goal, in general, we shouldn't be doing weird things just to accommodate the current coding, either. But is this weird enough to cross the line?
4. I really can't convince myself whether we need that third "only published data that I own" state, although I can see some boundary use cases for it.

Reply:
I would resist coding the Server to the expectations of the Dashboard:  while the Dashboard is currently our most important client, it might not be our only client, and a rational, well-grounded API design is probably more important than compatibility, at this point.

So, the question is, should the server enable the client to fetch exactly the data that it wants in a single request, or is it OK (even good!) to require the client to make multiple requests for complicated cases?

Off the top of my head, I can imagine users wanting the following sorts of collections:

- All my data
- All my data which is private
- All my data which is published
- All of User-X's data (but, unless I'm privileged, I'll only get to see published data)
- All of User-X's data which is published
- Any user's data which is published

Now, the first three could be collapsed into a single request, if we enable and require the client to filter the results, but I'm thinking that that might be onerous and complicated (the next two could be similarly collapsed, if we wanted to do that).  We can combine the "my" vs. "other user" requests by requiring the "my data" request to specify "my username"; and, we can combine the "all" request with them by coming up with some way to indicate "all" instead of a specific user.

So, one way to proceed would be to make the `"user"` parameter specify the user to select, unless it is "falsey", in which case it means "all users" (I'm disinclined to distinguish between "omitted", `None`, and `""` -- for simplicity I would make them all mean "all"), and then add another selector for "public", "private", or "both" (e.g., using `"access": "public"`, `"access": "private"`, or "omitted", meaning "both").  I believe that this allows a single request to fetch: all data for a single user (including self) or for all users, intersected with public, private, or both, with the simplest request (i.e., with `"user"` and `"access"` omitted) getting all possible data.  If the user wants a conjunction such as "my data plus public data", it would require two requests (plus de-duplication), although, unless the user is privileged, simply asking for _all_ data (i.e., omitting `"user"` and `"access"`) would produce the same result in that case.

Reply:
I liked the idea of defaulting a missing `"user"` to the authorized user; however you're correct that, in terms of API expressiveness, defaulting to "all users" is probably more useful. It also has the distinct advantage of making a missing `"user"` do the same thing regardless of whether you're logged in or not.

OK, so:

`{}` means all private and public data for all users (with the proviso that unless I'm ADMIN, I can only see my own private data).

`{"user": "drb"}` means all public and private data for user "drb".

`{"user": "drb", "access": "private"}` means only private data for user "drb".

`{"user": "drb", "access": "public"}` means only public data for user "drb".

`{"user": "drb"}` means all data owned by user "drb" both private and public.

`{"access": "public"}` means all public data regardless of the owner. (Which, for a non-logged-in user, is indistinguishable from `{}`, but would be different for a logged-in user, and more differenter for a logged-in ADMIN user.)

Reply:
Yup.

Only, you have one case in there twice and you're missing a case:

> {"user": "drb"} means all data owned by user "drb" both private and public.

should be `{"access": "private"}`, which means only private data for all users (which is a rather weird and unlikely case, but it might be one that is interesting -- i.e., it's a shorthand for "my (private) data" unless I'm an Admin, in which case, I'm less clear on what it might be used for...).

Reply:
> should be `{"access": "private"}`, which means only private data for all users (which is a rather weird and unlikely case, but it might be one that is interesting -- i.e., it's a shorthand for "my (private) data" unless I'm an Admin, in which case, I'm less clear on what it might be used for...).

Well, dang, you're correct; how did that happen?! I can't even say I copied-and-pasted and forgot to edit, because the wording is entirely different. ;-)